# Q1.
VALUES are 'hello' , -87.8 and 6.    

   EXPRESSIONS are * , - , / , + .

# Q2.
STRINGS -

   As far as python is concerned, anything you assign to a variable using quotes(single or double) will be considered as a string. Thus, a string is one of the several types of variable.

In [1]:
a='10'
b='20'
print(a+b)
type(a)

1020


str

In [2]:
a=10
b=20
print(a+b)
type(a)

30


int

So,in the above example (a+b) returns 1020 in the first case (string concatenation of 10 and 20),whereas in the second case, it returns 30 since it inteprets a and b as integers at run time.

VARIABLES -  

We can think of variable as an identifier(or some kind of means) to store a value in a memory location and to manipulate it if needed.

x=10               

x=20

So in the above example, we can say that x is a variable that stores the value 10. But, in Python there's no specific command to declare a variable. Therefore, a variable is created when you first assign a value to it in a program, the variable x is created when you assign the value 10 to it. Later when we re-assign x=20, the value in the corresponding memory location will be changed to 20. Thus, we can think of variable as an identifier to play around with a value that is created and stored in the memory location. A variable can be of several types like integer,string,float,etc based on the value we assign to it. 

# Q3.
DATA TYPES IN PYTHON ARE AS FOLLOWS    

NUMERIC -     

In Python, numeric data type represent the data which has numeric value. Numeric value can be integer, floating number or even complex numbers. These values are defined as int, float and complex class in Python.

In [3]:
#NUMERIC EXAMPLES
num =1      #integer number
type(num)

int

In [4]:
num1= 0.1    #float
type(num1)

float

In [5]:
num2= 6+7j    #complex
type(num2)

complex

SEQUENCE TYPE -   

In Python, sequence is the ordered collection of similar or different data types. Sequences allows to store multiple values in an organized and efficient fashion. There are several sequence types in Python –STRING,LIST AND TUPLE.

In [6]:
#SEQUENCE EXAMPLES

a= 'Alisha'    #STRING
type(a)

str

In [7]:
a=[2,5,7,9.0,'ALISHA']     #LIST
type(a)

list

In [8]:
a=(2,3,5,'DC')        #TUPLE
type(a)

tuple

BOOLEAN -   

TRUE and FALSE data is known as BOOLEAN DATA and the dat types which stores this boolean data is known as BOOLEAN DATA TYPES
#True and False with capital ‘T’ and ‘F’ are valid booleans otherwise python will throw an error.

In [9]:
#BOOLEAN EXAMPLES
a=True
type(a)

bool

In [10]:
b=False
type(b)

bool

SET -    

 Set is an unordered collection of data type that is iterable, mutable and has no duplicate elements.Sets can be created by using the built-in set() function with an iterable object or a sequence by placing the sequence inside curly braces, separated by ‘comma’. Type of elements in a set need not be the same, various mixed-up data type values can also be passed to the set.

In [11]:
#EXAMPLES OF SET
set1 = set([90,'dchouse',3,5]) 
print(set1) 

{90, 3, 5, 'dchouse'}


In [12]:
set1 = set("dc house") 
print(set1)

{'h', 'u', 'o', 'e', 'c', 's', ' ', 'd'}


DICTIONARY -    
 
 The order of elements in a dictionary is undefined. But we can iterate ovwer the following.
 1. The keys.
 2. The values.
 3. The items(key-value pairs) in a dictionary.
 
 A python dictionary is an unordered collection of key-value pairs. When we have large amount of data, the dictionary data type is used. Keys and values can be of any type in a dictionary. Items in dictionary are enclosed in the curly braces{} and separated by the comma(,). A colon (:) is used to separate key from value. A Key inside the square bracket[] is used for acessing the dictionary items.
 
 #type() function is used to determine the type of data type.

In [13]:
#EXAMPLE Of DICTIONARY
dict1={1:'first','second':2}    #declare dictionary
dict1[3]='third'                #add new item
dict1                           #display dictionary

{1: 'first', 'second': 2, 3: 'third'}

In [14]:
dict1.keys()       #display dictionary keys

dict_keys([1, 'second', 3])

In [15]:
dict1.values()        #display dictionary values

dict_values(['first', 2, 'third'])

Q4.   An EXPRESSION is a combination of variables,operators,values and reverse keyword. Whenever we type an expression in the command line, the interpreter evaluates it and produces the result.

An EXPRESSION is any sequence of literals (strings, numbers, lists, sets, tuples), objects, function calls, identifiers combined with a variety of operators which results in a value - regardless of whether of the resulting value is an object, or a number, or even a None. So examples of expressions are :

‘Hello World’ - this is a string literal.                        
23.6 - this is a floating point literal.                            
{1,2,3,4} - a set literal.                          
my_func(1,2,3) - a function call (even if the function returns None).                       
2 + the_number * math.factorial(7) - a combination of literals, operators, identifiers and function calls.           
EXPRESSIONS aren’t things like for loops, while loops, with statements, try/except, if/else - they are statements.

EXPRESSIONS are things that are intended to result in a value.


Operators are special symbols that designate that some sort of computation should be performed. The values that an operator acts on are called operands.

In [16]:
a=5
b=10
a+b

15

In this case, the + operator adds the operands a and b together. An operand can be either a literal value or a variable that references an object

In [17]:
a = 5
b = 10
a + b - 5

10

A sequence of operands and operators, like a + b - 5, is called an expression. Python supports many operators for combining data objects into expressions. These are explored below.

ARITHMETIC OPERATORS - These operators are used to perform arithmetic operations such as addition, subtraction, multiplication and division.

In [18]:
#EXAMPLES

x=10
y=20

In [19]:
z=x+y      #gives addition of two operands
print(z)

30


In [20]:
z=x-y
print(z)  #gives subtraction of two operands

-10


In [17]:
z=x*y
print(z)  #gives multiplication of two operands

200


In [18]:
z=x/y     #gives division of two operands
print(z)

0.5


In [19]:
z=x**y
print(z)  #exponential operator calculates power

100000000000000000000


In [20]:
z=x%y
print(z)  #modulus operator finds remainder

10


In [21]:
z=x//y
print(z)  #floor division operator gives quotient

0


COMPARISON OPERATORS - Comparison operators are also called as relational operators. The result of these operators is always Boolean value, that is, either TRUE or FALSE.

In [21]:
#EXAMPLES
x=10
y=12

In [22]:
if x==y:                   #(==) operator to check if two operands are equal
    print('x is equal to y')
else:
    print('x is not equal to y')

x is not equal to y


In [ ]:
if x!=y:                #(!=) operator to check if two operands are not equal.
    print('x is not equal to y')
else:
    print('x is equal to y')

In [13]:
if x<y:                   #(<) operator to check if first operand is smaller than the second operand.
    print('x is less than y')
else:
    print('x is not less than y')

x is less than y


In [15]:
if x>y:                   #(>) operator to check if first operand is greater than the second operand.
    print('x is greater than y')
else:
    print('x is not greater than y')

x is not greater than y


In [14]:
if x>=y:                   #(>=) operator to check if first operand is greater than or equal to second operand.
    print('x is greater than or equal to y')
else:
    print('x is not greater than or equal to y')

x is not greater than or equal to y


In [16]:
if x<=y:                   #(<=) operator to check if first operand is smaller than or equal to second operand.
    print('x is equal to or less than y')
else:
    print('x is not equal to or less than y')

x is equal to or less than y


ASSIGNMENT OPERATOR - Assignment operator is used to store right side operand in the left side operand.

In [57]:
#EXAMPLES
x=10
y=12

In [58]:
y+=x            #add right side operand to left side operand and store the result in left side operand .
print(y)

22


In [59]:
y*=x            #multiply right side operand with left side operand and store the result in left side operand .
print(y)

220


In [60]:
y/=x            #divide left side operand by right side operand and store the result in left side operand .
print(y)

22.0


In [61]:
y%=x            #find the modulus and store the remainder in the left side operand
print(y)

2.0


In [62]:
y**=x           #find exponential and store the result in the left side operand.
print(y)

1024.0


In [63]:
y//=x           #find the floor division and store in left side operand.
print(y)

102.0


LOGICAL OPERATORS - Logical operators are used to check two or more conditions.
The resultant of this operator is always BOOLEAN value.

In [66]:
#EXAMPLES
x= True
y= False

In [71]:
print(x and y)         #(AND) When both operands are true resultant becomes true

False


In [72]:
print(x or y)     #(OR) When any operand is true resultant becomes true.

True


In [73]:
print(not x)            #NOT operator is used to reverse the operand state.

False


In [75]:
print(not y)

True


MEMBERSHIP OPERATORS - These operators are used to check an item or an element that is part of a string, a list or a tuple. A membership operator reduces the effort of searching an element in the list. Suppose, x stores a value 20 and y is in the list containing items 10,20,30 and 40. Then, x is part of the list y because the value 20 is in the list y.

In [77]:
#EXAMPLES
x= 10
y= 20
list = [21,13,10,17]

In [80]:
if x in list:              #(IN) returns true if item is in the list of the sequence.
    print('x is present in the list')
else:
    print('x is not present in the list')

x is present in the list


In [82]:
if y not in list:            #(NOT IN) returns true if item is not in the list of the sequence.
    print('y is not present in the list')
else:
    print('y is present in the list')

y is not present in the list


IDENTITY OPERATORS - These operators are used to check whether both operators are same or not.

In [83]:
#EXAMPLES
x=12
y=12
if x is y:           #(Is) Returns true if operands are same.
    print('x is same as y')
else:
    print('x is not same as y')

x is same as y


In [84]:
y=20
if x is not y :     #(IS NOT) Returns false if operands are same.
    print('x is not same as y')
else:
    print('x is same as y')

x is not same as y


BITWISE OPERATORS - These operators perform bit level operation on operands. Lets us take two operands x=10 and y = 12. In, BINARY format this can be written as x = 1010 and y =1100.

In [90]:
#EXAMPLES
x=10          #10 = 00001010
y=12          #12 = 00001100

In [91]:
#BITWISE AND
x & y          # & operator copies bit if it exists in both the operands.

8

In [92]:
#BITWISE OR
x|y            # | operartor copies bit if it exists in either operands. 

14

In [93]:
#BITWISE XOR
x^y              # ^ operator copies bit if it exists in only one operand.

6

In [94]:
#BITWISE INVERSE
~x           #~ operator is a uninary operator used t opposite the bits of the operand.

-11

In [96]:
#LEFT SHIFT
x<<2        # << operator shifts the bit towards left.

40

In [97]:
#RIGHT SHIFT
x>>2       #>> operator is used to shift the bit towards right.

2

# Q5. 
A statement is an instruction that the Python interpreter can execute. A statement is interpreted by the interpreter and after execution displays some result(if there is a need for displaying it). For example., Print statement produces some result to display but it doesn't happen in case of assignment statement. 

A program can contain many statements in a sequence. If there are multiple statements, the result is displayed after every statement.

Some examples Of Statements are while statement, for statement, if statement, print statement,etc. 

Example of Assignment statement - 

In [22]:
spam=10

Due to assignment operator (=), the message variable stores the 10.

An expression is a combination of values, variables, operators, and calls to functions. Expressions need to be evaluated. If you ask Python to print an expression, the interpreter evaluates the expression and displays the result.

In [23]:
#Example
print(2)

2


The evaluation of an expression produces a value, which is why expressions can appear on the right hand side of assignment statements. A value all by itself is a simple expression, and so is a variable. Evaluating a variable gives the value that the variable refers to.

In [32]:
y = 3
x = len("world")
print(x)
print(y)

5
3


 we will see one of the distinct differences between statements and expressions.

In [35]:
y = 3
x = len("world")
print(x)
print(y)

5
3


In [36]:
y

3

Note that when we enter the assignment statement, y = 3, only the prompt is returned. There is no value. This is due to the fact that statements, such as the assignment statement, do not return a value. They are simply executed.

On the other hand, the result of executing the assignment statement is the creation of a reference from a variable, y, to a value, 3. When we execute the print function working on y, we see the value that y is referring to. In fact, evaluating y by itself results in the same response.

# Q6

In [54]:
bacon = 22
bacon+1

23

In [55]:
bacon

22

Variable Bacon will contain 22 as value. The bacon+1 expression does not reassign the value in bacon, that would need an assignment statement:bacon= bacon+1

In [56]:
bacon =bacon +1
bacon

23

# Q7

In [57]:
'spam'+'spamspam'

'spamspamspam'

In [58]:
'spam'*3

'spamspamspam'

BOTH EXPRESSIONS EVALUATE THE STRING 'SPAMSPAMSPAM'

# Q8.

VARIABLES NAME CANNOT BEGIN WITH A NUMBER.

# Q9.
The int(),float() and str() are the three functions that can be used to get the integer,floating point number and string versions of value can be passed to them.

# Q10.

In [60]:
'I have eaten' + 99 + 'buritos'          #WE WILL GET AN ERROR. 

TypeError: can only concatenate str (not "int") to str

The expression causes an error because 99 is an integer and onlhy string can be concatenated to other strings with the '+' operator. 

In [24]:
#RIGHT WAY
'I have eaten' + str(99) +  'burritos'

'I have eaten99burritos'